astacus/node/sqlite_snapshot.py

```diff
@@ -0,0 +1,248 @@
+                mtime_ns integer not null,
+                hexdigest text not null,
+                content_b64 text,
+                primary key (relative_path)
+            );
+            create index snapshot_files_hexdigest on snapshot_files(hexdigest);
+            commit;
+            """
+        )
+        super().__init__(dst)
+
+    def __len__(self) -> int:
+        return self._con.execute("select count(*) from snapshot_files;").fetchone()[0]
+
+    def get_file(self, relative_path: Path) -> SnapshotFile | None:
+        cur = self._con.execute("select * from snapshot_files where relative_path = ?;", (str(relative_path),))
+        row = cur.fetchone()
+        return row_to_snapshotfile(row) if row else None
+
+    def get_files_for_digest(self, hexdigest: str) -> Iterable[SnapshotFile]:
+        return map(
+            row_to_snapshotfile,
+            self._con.execute(
+                """
+                select *
+                from snapshot_files
+                where hexdigest = ?
+                order by relative_path;
+                """,
+                (hexdigest,),
+            ),
+        )
+
+    def get_all_files(self) -> Iterable[SnapshotFile]:
+        return map(row_to_snapshotfile, self._con.execute("select * from snapshot_files order by relative_path;"))
+
+    @override
+    def get_all_paths(self) -> Iterable[Path]:
+        return (
+            Path(row[0]) for row in self._con.execute("select relative_path from snapshot_files order by relative_path;")
+        )
+
+    @override
+    def get_total_size(self) -> int:
+        return self._con.execute("select sum(file_size) from snapshot_files;").fetchone()[0] or 0
+
+    def get_connection(self) -> sqlite3.Connection:
+        return self._con
+
+    def get_all_digests(self) -> Iterable[SnapshotHash]:
+        for hexdigest, file_size in self._con.execute(
+            """
+            select hexdigest, file_size
+            from snapshot_files
+            where hexdigest != ''
+            order by hexdigest;
+            """
+        ):
+            yield SnapshotHash(hexdigest=hexdigest, size=file_size)
+
+
+class SQLiteSnapshotter(Snapshotter[SQLiteSnapshot]):
+    def __init__(
+        self, groups: Sequence[SnapshotGroup], src: Path, dst: Path, snapshot: SQLiteSnapshot, parallel: int
+    ) -> None:
+        super().__init__(groups, src, dst, snapshot, parallel)
+        self._con = snapshot.get_connection()
+
+    def perform_snapshot(self, *, progress: Progress) -> None:
+        files = self._list_files_and_create_directories()
+        new_or_existing = self._compare_current_snapshot(files)
+        for_upsert = self._compare_with_src(new_or_existing)
+        with_digests = self._compute_digests(for_upsert)
+        self._upsert_files(with_digests)
+        self._con.execute("drop table if exists new_files;")
+        self._con.commit()
+
+    def _list_files_and_create_directories(self) -> Iterable[Path]:
```

Comment:
This functionality feels odd to have duplicated (with somewhat different implementations) here and in memory snapshotter.

Reply:
Everything here is an iterator to not mess up the memory usage.  The original implementation adapted here would yield the same value twice if there are overlapping globs.  But the original implementation is probably also faster for very selective globs, since it might not need to recurse the whole tree (although I haven't checked this much).

Reply:
Yeah, I don't really care which implementation is picked, just that there's one (as opposed to two) as I don't really see benefit in having both.